src/adapter/src/coord/read_policy.rs

```diff
@@ -419,6 +420,23 @@ impl crate::coord::Coordinator {
         read_holds
     }
 
+    /// Attempt to acquire read holds on the indicated collections at the indicated `time`.
+    /// This is similar to [acquire_read_holds], but instead of returning the read holds,
+    /// it arranges for them to be automatically released at the end of the transaction.
+    ///
+    /// If we are unable to acquire a read hold at the provided `time` for a specific id, then we
+    /// will acquire a read hold at the lowest possible time for that id.
+    pub(crate) fn acquire_read_holds_auto_cleanup(
+        &mut self,
+        session: &mut Session,
+        time: Timestamp,
+        id_bundle: &CollectionIdBundle,
+    ) {
+        let read_holds = self.acquire_read_holds(time, id_bundle);
+        self.txn_read_holds
+            .insert(session.conn_id().clone(), read_holds);
```

Comment:
The insert could delete some existing read holds. This should instead do something like `.entry(connid).or_insert_with(|| ReadHolds::new()).extend(read_holds)`

Reply:
Oh, yes, sorry!

Actually, I copied this line from [here](https://github.com/MaterializeInc/materialize/blob/d6a3c785559d40564f9b70fb8536a43ac1e8cdc0/src/adapter/src/coord/sequencer/inner/peek.rs#L694). So then this is probably a bug also on `main`?

I've fixed both occurrences now.